### ISSUE TITLE
Add support for traits.

### ISSUE THREAD

**Issue by wapmorgan** [phalcon/zephir#504]
Hey. Hey.
Traits aren't supported yet?

**Comment by SneakyBobito** [comment]
Trait is an usefull and powerfull feature, though it's only compatible with 5.4 and latter, it mays be used.

And we cant just merge trait method in the class that use it at the compilation because we wont be able to use the trait from the php code.

**Comment by jimthunderbird** [comment]
I tend to agree with what fezfez said. We might want to stabilize zephir before adding new features. With php parser and abstract syntax tree, we will always be able to merge the code in trait into individual classes.

**Comment by SneakyBobito** [comment]
By only merging traits into classes, you wlil break the API. Traits are not used only inside the library, they are also usable outside a library. Better no implement them rather than merge methods.

**Comment by hylent** [comment]
How about giving up PHP5.3 support?

**Comment by SneakyBobito** [comment]
Trait are definitively not supported in php 5.3, right ?

**Comment by hylent** [comment]
I mean Zephir giving up support for PHP 5.3.
There are problems when running Zephir extensions on PHP 5.3, which I can hardly trace.
By giving up support for PHP 5.3, Traits are reasonable feature for Zephir.

**Comment by jimthunderbird** [comment]
In my tool PHP-To-C-Ext, it now allows using trait to build zephir based php extensions. Under the hood it just merge all functions in the trait into individual classes.

https://github.com/jimthunderbird/php-to-c-extension#example-07

**Comment by SneakyBobito** [comment]
Traits are not aimed to be merged into class.
If you want zephir to support this feature , just create another type specific to zephir, and call it mixin for instance.

A php class that uses a trait is NOT compatible with php 5.3 and so should go for zephir.
You should see traits as a part of the public API, not as something used internaly in the library.

Sometimes traits are even not used internaly in the library, they are just utilities available for everyone that use the library.

Believe me or not, but merging traits into classes and throwing them away from the final extension will bring a lot of confusions.

**Comment by Nirlah** [comment]
+1

**Comment by sergeyklay** [comment]
Judging by this infographic http://php.net/supported-versions.php Soon we will lose the support of 5.4. Maybe it's time to use Traits? :)

**Comment by jimthunderbird** [comment]
It would be nice if zephir support trait as a built-in structure, if not, right now I am still going the method merging ways, developers have options not to use trait.
https://github.com/jimthunderbird/php-to-c-extension

**Comment by aurimasniekis** [comment]
Any update on this?

**Comment by flip111** [comment]
+1

**Comment by Green-Cat** [comment]
Should be possible now that support for php 5.3 is dropped.

**Comment by baohanddd** [comment]
Hope implement traits soon +1

**Comment by sergeyklay** [comment]
@Green-Cat +1

**Comment by aschwin** [comment]
I figure there are three ways to use traits within Zephir:
1. As Zephir traits to add functionality for Zephir classes
2. As PHP traits to add functionality for PHP classes
3. As a combination of both

I think no.3 is powerful, but confusing. I think no.1 is nice, but I think most useful would be no.2? Best would be both no.1 and no.2, where 1 trait is not both for Zephir and PHP.

Like suggested it would be good to make them distinct: mixin for Zephir (no.1), trait for PHP (no.2). No.3 would be implicit be no.1 and no.2. That is confusing. I think if functionality is shared, a no.2 that implements no.1 for example would be better maintainable.

**Comment by flip111** [comment]
@aschwin zephir is like a layer on top of C with zend Api. Imo it would be best to make it as transparent as possible so that a class in zephir is a class in php and a trait in zephir a trait in php. If you want to _hide_ stuff from php it would be best to use a technique like optimizers. This technique could probably be extended with more funtionality (on the part of the zephir programmer), but implementing this can be much harder. For example if you want to hide a zephir class from php, because all the classes are registered with the zend api, it would require custom logic so that they can not be used in php.

**Comment by gsouf** [comment]
@flip111 is problably one of the few to understand what a is a trait: a trait **is** a trait.

**Comment by Izopi4a** [comment]
+1

**Comment by gtux** [comment]
+1

**Comment by kuzmichus** [comment]
+1

**Comment by Jurigag** [comment]
+1

**Comment by Jurigag** [comment]
I will try to add trait support for PHP only, support for zephir itself will be too hard for me i guess.

**Comment by Jurigag** [comment]
But the problem with traits is that internal traits started working just from latest php 7.0.12 for php 7, didn't check php 5 yet.

So currently if traits would be added then in php 7 when using in zephir they would work for 7.0.0, but using them in php is working from 7.0.12

**Comment by ZanderBrown** [comment]
2018 Calling

**Comment by wapmorgan** [comment]
More likely we will see php 8 rather than traits

**Comment by marcus-sa** [comment]
I feel like this language lacks... everything

**Comment by Jurigag** [comment]
Because it was mainly done for writing phalcon and its features were driven by phalcon development.

**Comment by dschissler** [comment]
It would be really amazing to have Traits for the Phalcon 4 era.  Not having them is really affecting the design of my extension.

**Comment by Ilhampasya** [comment]
i hope phalcon need trait, so we can use trait on zephir haha
